[CT-1813] [CT-1378] [Bug] Python models not picking up custom schema

### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When I try to create a python model using v1.3, I get an error that a schema was not provided. I have set a schema in my `dbt_project.yml` file for the directory that the model is in, and have the `generate_schema_name` set to change the schema based on the environment.

### Expected Behavior

The python model will run and create a model in the same schema as sql models in the same directory

### Steps To Reproduce

1. Set the `dbt_project.yml` to a use a given schema, e.g.
```
models:
  dbt:
    datamarts:
      dim:
        schema: dim
```
2. add a python file `dbt/models/datamarts/dim/test_python.py` in that directory, e.g:
```
def model(dbt, session):
    dbt.config(materialized="table")
    df = dbt.ref("my_other_model").to_pandas()
    return df
```
3. create the macros `generate_schema_name_for_env` and `generate_schema_name`:
```
{% macro generate_schema_name_for_env(custom_schema_name, node) -%}
    {%- set default_schema = target.schema -%}
    {%- if target.name = 'prod' and custom_schema_name is not none -%}
        {{ custom_schema_name | trim }}
    {%- else -%}
        {{ default_schema }}_{{custom_schema_name | trim}}
    {%- endif -%}
{%- endmacro %}


{% macro generate_schema_name(custom_schema_name, node) -%}
    {{ generate_schema_name_for_env(custom_schema_name, node) }}
{%- endmacro %}
```
4. Run the python model

### Relevant log output

```shell
"/usr/lib/python_udf/431102875043084cdb1968114cbff94386b72d5e184e97acf96b1780c382f509/lib/python3.8/site-packages/snowflake/connector/pandas_tools.py", line 122, in write_pandas
      raise ProgrammingError(
  snowflake.connector.errors.ProgrammingError: Schema has to be provided to write_pandas when a database is provided
   in function TEST_PYTHON__DBT_SP with handler main
  compiled Code at target/run/dbt/models/datamarts/dim/test_python.py
```


### Environment

```markdown
- OS: dbt Cloud IDE
- Python: python in dbt cloud IDE
- dbt: 1.3
```


### Which database adapter are you using with dbt?

snowflake

### Additional Context

_No response_

I am also on:
```
dbt-core 1.4.5
dbt-snowflake 1.4.2
snowflake-connector-python 3.0.0
```

My custom schema is DEV_MODEL_python_model_test
The default schema is ADMIN

My python model:
`def model(dbt, session):
    """
    Entry point for dbt
    """
    dbt.config(materialized = "table")
    df = dbt.ref("users")
    pandas_df = df.limit(100).to_pandas()
    return pandas_df`

There error:
`02:10:48  Database Error in model python_model (models/model/python_model_test/python_model.py)
02:10:48    100357 (P0000): Python Interpreter Error:
02:10:48    Traceback (most recent call last):
02:10:48      File "_udf_code.py", line 94, in main
02:10:48      File "_udf_code.py", line 88, in materialize
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/snowpark/session.py", line 1453, in create_dataframe
02:10:48        t = self.write_pandas(
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/snowpark/session.py", line 1367, in write_pandas
02:10:48        raise pe
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/snowpark/session.py", line 1346, in write_pandas
02:10:48        success, nchunks, nrows, ci_output = write_pandas(
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/pandas_tools.py", line 223, in write_pandas
02:10:48        cursor.execute(file_format_sql, _is_internal=True)
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/cursor.py", line 829, in execute
02:10:48        Error.errorhandler_wrapper(
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/errors.py", line 232, in errorhandler_wrapper
02:10:48        handed_over = Error.hand_to_other_handler(
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/errors.py", line 287, in hand_to_other_handler
02:10:48        cursor.errorhandler(connection, cursor, error_class, error_value)
02:10:48      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/errors.py", line 165, in default_errorhandler
02:10:48        raise error_class(
02:10:48    snowflake.connector.errors.ProgrammingError: 003001 (42501): SQL access control error:
02:10:48    Insufficient privileges to operate on schema 'ADMIN'
02:10:48     in function PYTHON_MODEL__DBT_SP with handler main
02:10:48    compiled Code at target/run/consumer_dbt/models/model/python_model_test/python_model.py
02:10:48  
02:10:48  Finished running 1 table model in 0 hours 0 minutes and 24.04 seconds (24.04s).
02:10:48  Encountered an error:
FailFast Error in model python_model (models/model/python_model_test/python_model.py)
  Failing early due to test failure or runtime error
Error: Process completed with exit code 2.`

I've tried to write the pandas_df to a snowpark df by running write_pandas e.g.:
`def model(dbt, session):
    """
    Entry point for dbt
    """
    dbt.config(materialized = "table")
    df = dbt.ref("participant_characteristics_wide")
    pandas_df = df.limit(100).to_pandas()
    new_df = session.write_pandas(
        df = pandas_df,
        table_name = f'{dbt.this.identifier}_temp',
        database = dbt.this.database,
        schema = dbt.this.schema,
        auto_create_table = True,
        table_type = 'temp',
        overwrite = True
    )
    return new_df`

But then I get a new error which seems to have a syntax issue in the schema name (extra quotation marks):
`02:51:40  Database Error in model python_model (models/model/python_model_test/python_model.py)
02:51:40    100357 (P0000): Python Interpreter Error:
02:51:40    Traceback (most recent call last):
02:51:40      File "_udf_code.py", line 102, in main
02:51:40      File "_udf_code.py", line 16, in model
02:51:40      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/snowpark/session.py", line 1367, in write_pandas
02:51:40        raise pe
02:51:40      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/snowpark/session.py", line 1346, in write_pandas
02:51:40        success, nchunks, nrows, ci_output = write_pandas(
02:51:40      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/pandas_tools.py", line 202, in write_pandas
02:51:40        cursor.execute(drop_table_sql, _is_internal=True)
02:51:40      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/cursor.py", line 829, in execute
02:51:40        Error.errorhandler_wrapper(
02:51:40      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/errors.py", line 232, in errorhandler_wrapper
02:51:40        handed_over = Error.hand_to_other_handler(
02:51:40      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/errors.py", line 287, in hand_to_other_handler
02:51:40        cursor.errorhandler(connection, cursor, error_class, error_value)
02:51:40      File "/usr/lib/python_udf/694df3cf538f091a277e3df26e69af371a3630fe8656f8ad7d386069b297bf34/lib/python3.8/site-packages/snowflake/connector/errors.py", line 165, in default_errorhandler
02:51:40        raise error_class(
02:51:40    snowflake.connector.errors.ProgrammingError: 002003 (02000): SQL compilation error:
02:51:40    Schema 'STAGE."DEV_MODEL_python_model_test"' does not exist or not authorized.
02:51:40     in function PYTHON_MODEL__DBT_SP with handler main
02:51:40    compiled Code at target/run/consumer_dbt/models/model/python_model_test/python_model.py
02:51:40  
02:51:40  Finished running 1 table model in 0 hours 0 minutes and 23.87 seconds (23.87s).
02:51:40  Encountered an error:
FailFast Error in model python_model (models/model/python_model_test/python_model.py)
  Failing early due to test failure or runtime error
Error: Process completed with exit code 2.`

--- a/.changes/unreleased/Fixes-20230518-011334.yaml
+++ b/.changes/unreleased/Fixes-20230518-011334.yaml
@@ -1,0 +1,6 @@
+kind: Fixes
+body: Address some issues regarding gustom schema overrides.
+time: 2023-05-18T01:13:34.108819-07:00
+custom:
+  Author: versusfacit
+  Issue: "393"

--- a/dbt/include/snowflake/macros/materializations/table.sql
+++ b/dbt/include/snowflake/macros/materializations/table.sql
@@ -47,6 +47,8 @@ def materialize(session, df, target_relation):
     if importlib.util.find_spec(package_name):
         import pandas
         if isinstance(df, pandas.core.frame.DataFrame):
+          session.use_database(target_relation.database)
+          session.use_schema(target_relation.schema)
           # session.write_pandas does not have overwrite function
           df = session.createDataFrame(df)
     {% set target_relation_name = resolve_model_name(target_relation) %}

--- a/tests/functional/adapter/test_python_model.py
+++ b/tests/functional/adapter/test_python_model.py
@@ -6,6 +6,49 @@ from dbt.tests.adapter.python_model.test_python_model import (
 )
 
 
+models__simple_python_model = """
+import pandas
+
+def model(dbt, session):
+    dbt.config(
+        materialized='table',
+    )
+    data = [[1,2]] * 10
+    return pandas.DataFrame(data, columns=['test', 'test2'])
+"""
+
+models__simple_python_model_v2 = """
+import pandas
+
+def model(dbt, session):
+    dbt.config(
+        materialized='table',
+    )
+    data = [[1,2]] * 10
+    return pandas.DataFrame(data, columns=['test1', 'test3'])
+"""
+
+models__custom_target_model = """
+import pandas
+
+def model(dbt, session):
+    dbt.config(
+        materialized="table",
+        schema="MY_CUSTOM_SCHEMA",
+        alias="_TEST_PYTHON_MODEL",
+    )
+
+    df = pandas.DataFrame({
+        'City': ['Buenos Aires', 'Brasilia', 'Santiago', 'Bogota', 'Caracas'],
+        'Country': ['Argentina', 'Brazil', 'Chile', 'Colombia', 'Venezuela'],
+        'Latitude': [-34.58, -15.78, -33.45, 4.60, 10.48],
+        'Longitude': [-58.66, -47.91, -70.66, -74.08, -66.86]
+    })
+
+    return df
+"""
+
+
 class TestPythonModelSnowflake(BasePythonModelTests):
     pass
 
@@ -19,28 +62,6 @@ class TestIncrementalSnowflakeQuoting(BasePythonModelTests):
     @pytest.fixture(scope="class")
     def project_config_update(self):
         return {"quoting": {"identifier": True}}
-
-
-models__simple_python_model = """
-import pandas
-
-def model(dbt, session):
-    dbt.config(
-        materialized='table',
-    )
-    data = [[1,2]] * 10
-    return pandas.DataFrame(data, columns=['test', 'test2'])
-"""
-models__simple_python_model_v2 = """
-import pandas
-
-def model(dbt, session):
-    dbt.config(
-        materialized='table',
-    )
-    data = [[1,2]] * 10
-    return pandas.DataFrame(data, columns=['test1', 'test3'])
-"""
 
 
 class TestChangingSchemaSnowflake:
@@ -96,3 +117,15 @@ class TestImportSnowflake:
             f"PUT file://{project.project_root}/seeds/iris.csv @dbt_integration_test/;"
         )
         run_dbt(["run"])
+
+
+# https://github.com/dbt-labs/dbt-snowflake/issues/393 is notorious for being triggered on some
+# environments but not others. As of writing this, we don't know the true root cause. This test may
+# not fail on all systems.
+class TestCustomSchema:
+    @pytest.fixture(scope="class")
+    def models(self):
+        return {"custom_target_model.py": models__custom_target_model}
+
+    def test_custom_target(self, project):
+        run_dbt()

--- a/tests/functional/adapter/test_python_model.py
+++ b/tests/functional/adapter/test_python_model.py
@@ -5,7 +5,6 @@ from dbt.tests.adapter.python_model.test_python_model import (
     BasePythonIncrementalTests,
 )
 
-
 models__simple_python_model = """
 import pandas
 
@@ -121,11 +120,12 @@ class TestImportSnowflake:
 
 # https://github.com/dbt-labs/dbt-snowflake/issues/393 is notorious for being triggered on some
 # environments but not others. As of writing this, we don't know the true root cause. This test may
-# not fail on all systems.
-class TestCustomSchema:
+# not fail on all systems with problems regarding custom schema model configurations.
+class TestCustomSchemaWorks:
     @pytest.fixture(scope="class")
     def models(self):
         return {"custom_target_model.py": models__custom_target_model}
 
     def test_custom_target(self, project):
-        run_dbt()
+        results = run_dbt()
+        assert results[0].node.schema == f"{project.test_schema}_MY_CUSTOM_SCHEMA"

